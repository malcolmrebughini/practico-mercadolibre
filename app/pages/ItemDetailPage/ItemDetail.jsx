import React from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import TopBar from '../../containers/TopBar';
import Breadcrumbs from '../../containers/Breadcrumbs';
import ItemDetail from '../../containers/ItemDetail';


class ItemDetailPage extends React.Component {
  componentWillMount = () => {
    if (Object.keys(this.props.data).length === 0) this.props.getItem(this.props.itemId);
  }

  componentWillUnmount = () => {
    this.props.clearData();
  }

  render = () => {
    const categories = this.props.categories.reduce((string, value, index, array) => {
      if (array.length === index + 1) return `${string} ${value}.`;
      return `${string} ${value},`;
    }, '');
    const price = this.props.data.price;
    const description =
      `Cómpralo en Mercado Libre a $ ${price} - Encuentra más productos de ${categories}`;

    return (
      <div>
        <TopBar />
        { !this.props.isLoading &&
          <div>
            <Helmet>
              <title>{`${this.props.data.title} - $ ${price} en MercadoLibre`}</title>
              <meta name="description" content={description} />
            </Helmet>
            <Breadcrumbs />
            <ItemDetail />
          </div>
        }
      </div>
    );
  }
}


ItemDetailPage.propTypes = {
  data: PropTypes.shape({
    price: PropTypes.number,
    title: PropTypes.string,
  }),
  itemId: PropTypes.string,
  categories: PropTypes.arrayOf(PropTypes.string),
  isLoading: PropTypes.bool,
  getItem: PropTypes.func,
  clearData: PropTypes.func,
};

ItemDetailPage.defaultProps = {
  data: {},
};


export default ItemDetailPage;
