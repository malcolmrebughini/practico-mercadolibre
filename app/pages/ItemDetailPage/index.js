import { connect } from 'react-redux';
import actions from '../../actions';
import ItemDetailPage from './ItemDetailPage';


function mapStateToProps(state, ownProps) {
  return {
    state,
    isLoading: state.itemDetail.isLoading,
    data: state.itemDetail.data,
    itemId: ownProps.match.params.itemId,
    categories: state.categories,
  };
}


export default connect(mapStateToProps, actions)(ItemDetailPage);
