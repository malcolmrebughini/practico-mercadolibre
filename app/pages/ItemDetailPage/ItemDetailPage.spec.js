/* eslint-env jasmine */
/* eslint react/jsx-filename-extension: 0 */

import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ItemDetailPage from './ItemDetailPage';


describe('ItemDetailPage', () => {
  it('should render correctly', () => {
    const data = {
      price: { integer: 90000, decimals: 0 },
      title: 'Some title',
    };
    const wrapper = shallow(
      <ItemDetailPage
        data={data}
        isLoading={false}
        itemId="MLA660069628"
        categories={['Some', 'Categories']}
        getItem={jest.fn()}
        clearData={jest.fn()}
      />,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call getItem if no data', () => {
    const callback = jest.fn();
    shallow(<ItemDetailPage getItem={callback} itemId="MLA660069628" />);

    expect(callback.mock.calls.length).toEqual(1);
    expect(callback.mock.calls[0]).toEqual(['MLA660069628']);
  });

  it('should call clearData on unmount', () => {
    const callback = jest.fn();
    shallow(<ItemDetailPage getItem={jest.fn()} clearData={callback} />).unmount();

    expect(callback.mock.calls.length).toEqual(1);
  });
});
