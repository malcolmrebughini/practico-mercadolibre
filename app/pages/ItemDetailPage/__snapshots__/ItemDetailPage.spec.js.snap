// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetailPage should render correctly 1`] = `
<div>
  <Connect(TopBar) />
  <div>
    <HelmetWrapper
      encodeSpecialCharacters={true}
    >
      <title>
        Some title - $ 90,000 en MercadoLibre
      </title>
      <meta
        content="Cómpralo en Mercado Libre a $ 90,000 - Encuentra más productos de Some, Categories."
        name="description"
      />
    </HelmetWrapper>
    <Connect(Breadcrumbs) />
    <Connect(ItemDetail) />
  </div>
</div>
`;
